Added ability to copy a specific version of an app

orangejenny opened the pull request:
The first commit just cherry-picks https://github.com/dimagi/commcare-hq/pull/24928 (and https://github.com/dimagi/commcare-hq/pull/24943) so we can release this functionality separately, as discussed in meeting with QA earlier this week.

[QA ticket](https://dimagi-dev.atlassian.net/browse/QA-666)

@dimagi/product Are y'all okay with releasing this to all users?

<img width="557" alt="Screen Shot 2019-08-15 at 4 16 12 PM" src="https://user-images.githubusercontent.com/1486591/63123901-09a06c80-bf78-11e9-9aa0-3b1286cd5fcd.png">

orangejenny committed:
Added ability to copy a specific version of an app

orangejenny committed:
Renamed _unset_build_fields since it's called outside the class

orangejenny committed:
Combined VersionedDoc and ApplicationBase

orangejenny committed:
Added copy_of to *_build_fields methods

orangejenny commented on corehq/apps/app_manager/models.py:
```diff
@@ -3864,162 +3864,6 @@ def super_save():
             super_save()
 
 
-class VersionedDoc(LazyBlobDoc):
```
So I'm not totally in love with this change, since `ApplicationBase` is already 600 lines long, and a VersionedDoc is a relatively clear concept (although the ODK urls don't belong in it...). But I was finding it hard to work with (because VersionedDoc controls copy_of but ApplicationBase controls all the other build-related fields). I'd rather collapse the unused level of hierarchy and instead break ApplicationBase up using composition.

Open to feedback. @proteusvacuum @emord you've also thought about the app manager models hierarchy.

mkangia replied:
I would push for keeping `VersionedDoc`, retaining version related logic in it.

And for build related things to go into application or if possible, in a different class and then that is used to make build related things instead of populating `Application` class. Inheriting `Application` from it would be the easier way out but let's avoid that. I have extracted classes like `HostedCCZUtility` which dealt with creating the CCZ file for `HostedCCZ` and related operations (the build in this case for Application) and then use that to handle operations.

orangejenny replied:
Looked at this a bit more, and I still like the idea of making VersionedDoc a separate utility class, but doing that is more complicated than I expected. VersionedDoc isn't a clean abstraction, given its use of application-specific urls (short_odk_url, etc) and the `app_manager/applications` couch view.

I don't think it's worth the effort to clean up VersionedDoc enough to make it truly generic and pull it out of the Application hierarchy, and I still think the extra layer of hierarchy (e.g., having two implementations of `make_build`) adds more in complexity than it's worth. I'm inclined to leave this code as it is in this PR, removing VersionedDoc altogether.

orangejenny committed:
Placated Stickler

mkangia reviewed (commented):
not approving yet, would like to see where the conversation for `VersionedDoc` goes from few reviews.

orangejenny commented:
Note to self: remember to handle feedback from https://github.com/dimagi/commcare-hq/pull/24928#discussion_r307824546

orangejenny committed:
Renamed functions

esoergel commented on corehq/apps/app_manager/models.py:
```diff
@@ -4521,22 +4422,52 @@ def make_build(self, comment=None, user_id=None):
+        self.comment_from = user_id
+        self.is_released = False
+
+    def convert_build_to_app(self):
```
I really like the framing of `convert_app_to_build` and vice-versa, but I do notice that `_meta_fields` is used in `export_json` as a list of fields to be excluded.  I think it implements the same concept as `convert_build_to_app`.  If that's true, it'd be good to combine the two (maybe do the `_meta_fields` stuff here in `convert_build_to_app`, say)

orangejenny replied:
I don't think these are quite the same pieces of functionality, given that `VersionedDoc.make_build` excludes a similar but not identical list of properties: https://github.com/dimagi/commcare-hq/blob/dcaeefa1dd579d43b1eabe6795216c555f3d2eb5/corehq/apps/app_manager/models.py#L3912

esoergel replied:
Yeah, I see that as well.  Seems like there's some overlap between `convert_build_to_app`, `export_json`, and `from_source`, at least conceptually, but I do see there are differences too.  Not sure whether the differences are intentional/desirable or accidental.  I'll leave it to your judgement.

esoergel commented on corehq/apps/app_manager/views/apps.py:
```diff
@@ -393,26 +393,35 @@ def copy_app(request, domain):
         linked = data.get('linked')
         if linked:
-            return _create_linked_app(request, app, link_domain, data['name'])
+            return _create_linked_app(request, app, link_domain, data['name'], master_build_id=data['build_id'])
```
Shouldn't `app` here come from `app_id_or_source` instead?

orangejenny replied:
Good catch, this wasn't working for linked apps. Added a few commits to fix, clean up, and add a test.

orangejenny committed:
Fixed logic for copying a specific version into a linked app

orangejenny committed:
Cleaned up params for _create_linked_app

orangejenny committed:
fixup! Fixed logic for copying a specific version into a linked app

orangejenny committed:
Added test of updating linked app to a specific version

orangejenny committed:
Fixed lint error

esoergel commented on corehq/apps/app_manager/views/apps.py:
```diff
@@ -416,25 +417,23 @@ def _inner(request, link_domain, data, master_domain=domain):
-    if not master_version:
+def _create_linked_app(request, master_build, link_domain, link_app_name):
+    master_domain = master_build.domain
+    if not master_build:
```
Not sure what the correct approach is, but this needs some rethinking - if `master_build` is `None`, `master_build.domain` will fail above this, and `master_build.master_id` will fail below.

orangejenny replied:
Clearly I'm having trouble getting my head all the way around `copy_app`. I took a fresh look at it and refactored in https://github.com/dimagi/commcare-hq/pull/25003/commits/ad3fb0a32d62a62cec26eb949221f8aeba2a307e with two goals: 1) move as much logic as possible inside `_inner`, rather than having some logic before it and then some inside and 2) make the function take two separate paths (normal vs linked app) that don't share much logic.

orangejenny committed:
Cleaned up logic in copy_app

orangejenny committed:
Added patch of validate_xform

esoergel reviewed (approved):
Nice, thanks - I also had trouble wrapping my head around this functionality

orangejenny committed:
Merge branch 'master' into jls/copy-app-version

orangejenny committed:
Switched test_update_from_specific_build to use its own apps.

Otherwise there are issues with this test failing when run with the entire test suite.

orangejenny committed:
Merge branch 'master' into jls/copy-app-version

orangejenny commented:
Marking don't merge again until I update the labeling per discussion with product.

orangejenny committed:
Updated labeling

esoergel commented:
:tada:
